### ISSUE TITLE
Versioning or backwards compatibility in message assembly

### ISSUE THREAD

**Issue by davidzwa** [akkadotnet/akka.net#4328]
- Akka, Akka.Remote: Version 1.3.17
- Platform: Windows, .NET Core 3.1
- Question: 
**Im trying to figure out (on an architectural level) how to enforce a backwards compatibility in message types and their migrations/updates.**

Imagine, someone setup a cluster of two devices. One device has version 1.0 of our assembly running and the other has version 1.1 running with a breaking change on message level. How would we approach that the two devices can still communicate the message? Are there good ways to introduce f.e. compile time checks (is this even possible)?
I see two approaches:
- Implementing partial/full serialization with versioning (**runtime** semi-compatibility)
- Implementing message intermediary (**runtime** incompatibility checks)
- Some sort of interfacing construct or backwards compatibility mapping (**design debt**)

Currently I don't see how I can enforce a way for our developers to maintain compatibility without breaking remote association/deployment. Any thoughts, libraries or something I'm missing?

**Comment by Aaronontheweb** [comment]
From our yet-to-be-publicly-promoted Clustering + K8s course: https://petabridge.com/cluster/lesson3 

See "Using Google Protocol Buffers to Version State and Messages"

And here's some diagrams which from that training which illustrate how to do it: 

https://www.slideshare.net/petabridge/akkanet-best-practices-for-continuous-deployment-w-akkacluster

TL;DR; - the way to preserve wire format is to use "extend-only" design with your wire messages. Anything that existed in the past never gets changed, moved, or deleted. New things can be added, old things can be deprecated but not removed.

Please let us know if this helps!

**Comment by davidzwa** [comment]
This is quite helpful indeed and thank you for your rapid response!

**Comment by davidzwa** [comment]
@Aaronontheweb on hind-sight, you mentioned the deprecation policy. If one is forced to remove a property, deprecation is not enough. I'm just speculating about the implications of 'never removing' as it is quite inflexible.

Do you have any plans for this, except for accepting/swallowing errors?
